### PR TITLE
Allow entering era in date picker by typing

### PR DESCRIPTION
**Pull request by devongovett**
Depends on #3167.

This adds support for entering an era in DatePicker by typing the first letter. This enables users of mobile devices to select an era when up and down arrow keys are not available.

### CODE REVIEW

**Review by LFDanLu — CHANGES_REQUESTED** [review]
Just one error I noticed but otherwise confirmed that I can type for Japanese eras on Android and desktop. I don't see a era segment for the Ethiopic calendar though, is there supposed to be one?

**Comment by devongovett** [comment]
Yeah the focus issue is a known issue (came up in testing), which I will fix separately.

The way you found is the only way to enter a BC date. I don't figure it's super common so no need to display AD all the time IMO. Could add an option for it in the future if we had to for some reason.

File: packages/@react-aria/datepicker/src/useDateSegment.ts [hunk 1]
```diff
@@ -143,6 +144,30 @@ export function useDateSegment(segment: DateSegment, state: DateFieldState, ref:
...
   }, [amPmFormatter]);
 
+  // Get a list of formatted era names so users can type the first character to choose one.
+  let eraFormatter = useDateFormatter({era: 'narrow', timeZone: 'UTC'});
+  let eras = useMemo(() => {
+    let date = toCalendar(new CalendarDate(1, 1, 1), state.calendar);
+    let eras = state.calendar.getEras().map(era => {
+      let eraDate = date.set({year: 1, month: 1, day: 1, era}).toDate('UTC');
+      let parts = eraFormatter.formatToParts(eraDate);
+      let formatted = parts.find(p => p.type === 'era').value;
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-aria/datepicker/src/useDateSegment.ts:158]: Good catch. Fixed it. Seems like the era wasn't returned if the year was also omitted.

Also noticed that we were forcing a long era name in calendar for all non-gregorian calendars, so got rid of that. The era is shown automatically by the formatter only when needed.

**Review by snowystinger — APPROVED** [review]

**Review by majornista — CHANGES_REQUESTED** [review]
As with ArrowUp/ArrowDown to change era, we need to manage focus when changing the era removes the focused segment from the DateField. With Gregorian calendar, typing `A` with focus on `BC` will select the common era, but remove the segment causing focus to be lost.

**Comment by devongovett** [comment]
@majornista that should be solved by #3213 once both PRs are merged together.

**Review by majornista — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]
looks good to me for the most part, verified the other issues are now resolved. Just one thing I noticed with the Ethiopic era: once you select/confirm a date the era set is no longer changeable. Maybe something specific to how Ethiopic calendars work? Trying to dig into this, but not a blocker for this PR since this seems to be pre-existing behavior

**Comment by devongovett** [comment]
@LFDanLu interesting. I'll look at that separately. I think it's due to some weirdness with the Ethiopic calendar. The eras don't really work like in other calendars where there is a single timeline. Instead, they are two different reference points on the same timeline. Essentially, AA 5501 = AM 1. So when switching between eras I guess we need to also update the year field perhaps...

**Pull request merged**